packages/salesforce-adapter/src/transformers/reference_mapping.ts

```diff
@@ -0,0 +1,335 @@
+    serializationStrategy: 'relativeApiName',
+  },
+  {
+    src: { field: CPQ_LOOKUP_REQUIRED_FIELD, parentType: CPQ_PRICE_RULE },
+    serializationStrategy: 'relativeApiName',
+  },
+  {
+    src: { field: CPQ_LOOKUP_TYPE_FIELD, parentType: CPQ_PRICE_RULE },
+    serializationStrategy: 'relativeApiName',
+  },
+]
+
+const matchName = (fieldName: string, matcher: string | RegExp): boolean => (
+  _.isString(matcher)
+    ? matcher === fieldName
+    : matcher.test(fieldName)
+)
+
+const matchMetadataType = (elem: Element, type: string | undefined): boolean => (
+  // need typeName for backward compatibility - TODO remove after verifying
+  type === undefined ? true : (metadataType(elem) === type || elem.elemID.typeName === type)
+)
+
+export class FieldReferenceResolver {
+  src: ExactSourceDef | RegExSourceDef
+  serializationStrategy: ReferenceSerializationStrategy
+  target: ExtendedReferenceTargetDefinition | undefined
+
+  constructor(def: FieldReferenceDefinition) {
+    this.src = def.src
+    this.serializationStrategy = ReferenceSerializationStrategyLookup[
+      def.serializationStrategy ?? 'absoluteApiName'
+    ]
+    this.target = def.target
+      ? { ...def.target, lookup: this.serializationStrategy.lookup }
+      : undefined
+  }
+
+  static create(def: FieldReferenceDefinition): FieldReferenceResolver {
+    return new FieldReferenceResolver(def)
+  }
+
+  match(field: Field): boolean {
+    return (
+      matchName(field.name, this.src.field)
+      && matchMetadataType(field.parent, this.src.parentType)
+    )
+  }
+
+  field(): string | RegExp {
+    return this.src.field
+  }
+}
+
+export type ReferenceResolverFinder = (field: Field) => FieldReferenceResolver[]
+
+/**
+ * Generate a functioning filtering for relevant reference resolvers for a given field.
+ */
+export const generateReferenceResolverFinder = (
+  defs = fieldNameToTypeMappingDefs
+): ReferenceResolverFinder => {
+  const referenceDefinitions = defs.map(
+    def => FieldReferenceResolver.create(def)
+  )
+
+  const matchersByFieldName = _(referenceDefinitions)
+    .filter(def => _.isString(def.src.field))
+    .groupBy(def => def.src.field)
+    .value()
+  const regexFieldMatchersByParent = _(referenceDefinitions)
+    .filter(def => _.isRegExp(def.src.field) && _.isString(def.src.parentType))
+    .groupBy(def => def.src.parentType)
+    .value()
+
+  return (field => (
+    [
+      ...(matchersByFieldName[field.name] || []),
+      ...(regexFieldMatchersByParent[metadataType(field.parent)] || []),
+    ].filter(resolver => resolver.match(field))
+  ))
+}
+
+const getLookUpNameImpl = (defs = fieldNameToTypeMappingDefs): GetLookupNameFunc => {
+  const resolverFinder = generateReferenceResolverFinder(defs)
+
+  const determineLookupStrategy = (args: GetLookupNameFuncArgs): ReferenceSerializationStrategy => {
+    if (args.field === undefined) {
+      log.debug(`could not determine field for path ${args.path?.getFullName()}`)
```

Comment:
nit: log function accept format strings
```suggestion
      log.debug('could not determine field for path %s',  args.path?.getFullName())
```
(slightly more efficient because when the logging is turned of it won't have to create the string)
applies to a bunch of places